presto-main/src/main/java/com/facebook/presto/operator/MergeJoinSource.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.common.Page;
+import com.facebook.presto.memory.context.LocalMemoryContext;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+
+import java.io.Closeable;
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+import static com.facebook.presto.operator.Operator.NOT_BLOCKED;
+import static com.google.common.base.Preconditions.checkState;
+
+public class MergeJoinSource
+        implements Closeable
+{
+    private final Queue<Page> pages;
+    private final boolean buffered;
+    private LocalMemoryContext localMemoryContext;
+    private int pagesRemaining;
+    private boolean finishing;
+
+    // consumer is blocked initially, producer is not blocked
+    private ListenableFuture<?> consumerFuture = SettableFuture.create();
+    private ListenableFuture<?> producerFuture = NOT_BLOCKED;
```

Comment:
Do we need those two futures? can we just do: 

```
    public boolean isProducerBlocked() {
        if (bufferEnabled) {
            return false;
        }
        return pages.size() > 0;
    }

    public boolean isConsumerBlocked() {
        return pages.isEmpty();
    }
```

let me know if the above logic is insufficient

Reply:
we use the producer and consumer futures returned from MergeJoinSource in the isBlocked() methods for the left and right operators